esp/bindings/SOAP/xpp/fxpp/FragmentedXmlPullParser.h

```diff
@@ -0,0 +1,2145 @@
+            break;
+        }
+        fprintf(stdout, "\n");
+    }
+
+};
+
+/**
+ * IFragmentedXmlAssistant implementation supporting the registration of elements that may
+ * represent either external or embedded fragments.
+ * 
+ * For embedded fragments, the assistant recognizes which elements may contain embedded fragment
+ * data, which of those do contain embedded fragment data, and creates the fragment representing
+ * the embedded fragment data.
+ *
+ * For external fragments, the assistat recognizes which elements may contain external fragment
+ * references. For elements that may contain references, it relies upon a separate detector to
+ * determine if they do, in fact, contain an external reference. For elements that do contain
+ * external references, it relies on a separate loader to resolve the detected key(s) into a
+ * fragment with the requested data.
+ */
+class CFragmentedXmlAssistant : public CInterfaceOf<IFragmentedXmlAssistant>
+{
+public:
+    enum DetectedType
+    {
+        Unknown,
+        Resource,
+        Path,
+    };
+
+    /**
+     * Abstraction representing the "identity" of an external fragment. The identity of a fragment
+     * is the information located by a detector that can be used by a loader to get the fragment
+     * data. It is assumed that a fragment may be uniquely identified by at least one key and at
+     * most two keys. The first, or primary, key may be a file path or resource bundle identifier.
+     * The second, if used, may be an XPath referring to a nodeset within a broader XML document.
+     */
+    interface IExternalDetection : extends IInterface
+    {
+        virtual int         queryDetectedType() const = 0;
+        virtual const char* queryPrimaryKey() const = 0;
+        virtual const char* querySecondaryKey() const = 0;
+
+        inline bool isResourceType() const { return Resource == queryDetectedType(); }
+        inline bool isPathType() const     { return Path == queryDetectedType(); }
+    };
+
+    /**
+     * Abstraction describing a helper (an assistant's assistant) responsible for locating fragment
+     * identifiers within parser start tag structures. The detection of fragments is separated from
+     * the assistant to allow multiple implementations to support different markup language
+     * requirements.
+     */
+    interface IExternalDetector : extends IInterface
+    {
+        virtual IExternalDetection* detect(const StartTag& stag) const = 0;
+    };
+
+    /**
+     * Abstraction describing a helper responsible for loading detected fragments for an assistant.
+     * It is separate from the assistant to support alternate sources; a unit test might access in-
+     * memory data while other use cases might load files or perform resource bundle lookups.
+     */
+    interface IExternalLoader : extends IInterface
+    {
+        virtual IXmlFragment* load(const IExternalDetection& detected, XmlFragmentRules rules) const = 0;
```

Comment:
As we discussed in our meeting consider having the loader just call back into the calling object and passing the pointers and other info that currently goes into the IXmlFragment object.

For example (and names are just examples and feel free to change them):

 ```
   interface IFragementLocation : extends IInterface
   {
      virtual void addContent(const char *id, const char *content, unsigned length, , XmlFragmentRules rules) const = 0;
   };

   interface IExternalLoader : extends IInterface
   {
      virtual void load(IFragmentLocation &location, const IDetectedContent& detected, XmlFragmentRules rules) const = 0;
   };
```
As we discussed, to make this useful, you would have to figure out a way to make it efficient such as immediately create the XPP parser so it immediately has the copy of the data.